### PR TITLE
Add contributing document

### PR DESCRIPTION
**Pull request by juanfont**
This PR adds a `CONTRIBUTING.md` document, detailing headscale's contribution policy.

The document is open for discussion with the community, and we welcome suggestions and adjustments.

Our objective is to have a sustainable project that serves the community, and does not burn out its maintainer.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+This model has been chosen to reduce the risk of burnout by limiting the maintenance overhead of reviewing and validating third-party code.
+
+## Why do we have this model?
+
+Headscale has a small maintainer team that tries to balance working on the project, fixing bugs and reviewing contributions.
+
+When we work on issues ourselves, we develop first hand knowledge of the code and it makes it possible for us to maintain and own the code as the project develops.
+
+When code is contributed to the project, it is typically a positive thing. People enjoy and engage with our project, but it also comes with some challenges; we have to understand the code, we have to understand the feature, we might have to become familiar with external libraries or services that this new feature integrates with and it needs to be reviewed from a security perspective. And that is only when it comes to reviewing it. After the code has been merged, the feature has to be maintained, meaning that changes to external parts need to be updated, and kept working.
```
Comment (ohdearaugustin) [inline CONTRIBUTING.md:12]:
```suggestion
Code contributions are seen as a positive thing. People enjoy and engage with our project, but it also comes with some challenges; we have to understand the code, we have to understand the feature, we might have to become familiar with external libraries or services and we think about security implications. All those steps are required during reviewing process.  After the code has been merged, the feature has to be maintained. Any changes reliant on external services must be updated and expanded accordingly. .
```
A bit of rewording and shortening of sentences

**Review by ohdearaugustin — CHANGES_REQUESTED** [review]
I think this really clarifies many things. I just would rewords some parts of it. Especially the part that contributors must help maintain the features, because this part is a bit unrealistic.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+The review and the day-1 maintenance adds a significant burden on them maintainers. Often we hope that the contributor of the feature will help out, but we found that most of the time, they disappear when they have their new feature added.
+
+This means that when someone contributes, we are mostly happy about it, but we do have to run it through a series of checks to establish if we actually can maintain this feature.
+
+## What do we require?
+
+A general description is provided here and an explicit list is provided in our pull request template.
+
+All new features have to start with a design document that has to be discussed in the issue tracker (not discord) with a use case for the feature, how it can be implemented, who will implement it and what the plan for maintaining it is.
```
[the remark this answers is not in the document]
Comment (kradalby) [reply CONTRIBUTING.md:22]: must include usecase?

**Comment by juanfont** [comment]
@ohdearaugustin great suggestions :)

I will leave the PR open for a few days more, just in case someone would like to add something else.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+Headscale is "Open Source, acknowledged contribution", this means that any contribution will have to be discussed with the maintainers before being added to the project.
+This model has been chosen to reduce the risk of burnout by limiting the maintenance overhead of reviewing and validating third-party code.
+
+## Why do we have this model?
+
+Headscale has a small maintainer team that tries to balance working on the project, fixing bugs and reviewing contributions.
+
+When we work on issues ourselves, we develop first hand knowledge of the code and it makes it possible for us to maintain and own the code as the project develops.
+
+Code contributions are seen as a positive thing. People enjoy and engage with our project, but it also comes with some challenges; we have to understand the code, we have to understand the feature, we might have to become familiar with external libraries or services and we think about security implications. All those steps are required during reviewing process.  After the code has been merged, the feature has to be maintained. Any changes reliant on external services must be updated and expanded accordingly. .
```
Comment (ohdearaugustin) [inline CONTRIBUTING.md:12]:
```suggestion
Code contributions are seen as a positive thing. People enjoy and engage with our project, but it also comes with some challenges; we have to understand the code, we have to understand the feature, we might have to become familiar with external libraries or services and we think about security implications. All those steps are required during the reviewing process. After the code has been merged, the feature has to be maintained. Any changes reliant on external services must be updated and expanded accordingly. 
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+The review and the day-1 maintenance adds a significant burden on the maintainers. Often we hope that the contributor will help out, but we found that most of the time, they disappear after their new feature was added.
```
Comment (ohdearaugustin) [inline CONTRIBUTING.md:14]:
```suggestion
The review and day-1 maintenance adds a significant burden on the maintainers. Often we hope that the contributor will help out, but we found that most of the time, they disappear after their new feature was added.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+## What do we require?
+
+A general description is provided here and an explicit list is provided in our pull request template.
+
+All new features have to start out with a design document, which should be discussed on the issue tracker (not discord). It should include a use case for the feature, how it can be implemented, who will implement it and a plan for maintaining it.
+
+All features have to be end-to-end tested (integration tests) and have good unit test coverage to ensure that they work as expected. This will also ensure that the feature continues to work as expected over time. If a change cannot be tested, a strong case for why this is not possible needs to be presented.
+
+The contributor must help maintain the feature over time, if a feature is found to be left unmaintained, we will have to remove it.
```
Comment (ohdearaugustin) [inline CONTRIBUTING.md:26]:
```suggestion
The contributor should help to maintain the feature over time. In case the feature is not maintained probably, the maintainers reserve themselves the right to remove features they redeem as unmaintainable. This should help to improve the quality of the software and keep it in a maintainable state.
```

**Review by ohdearaugustin — CHANGES_REQUESTED** [review]
Just fixed some typos and added a new paragraph for the maintainability of features.

**Review by ohdearaugustin — APPROVED** [review]

**Comment by ohdearaugustin** [comment]
@juanfont still wanna leave it for some days or should I merge it?

**Comment by juanfont** [comment]
Merged :)
